composer/utils/collect_env.py

```diff
@@ -0,0 +1,391 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+"""Helpers to gather system information for debugging and bug reporting.
+
+Leverages PyTorch's :mod:`torch.utils.collect_env` package to gather pertinent system information.
+The following information is additionally collected to faciliate Comopser specific debug:
+
+* Composer version
+* Number of nodes
+* Host processor model name
+* Host processor physical core count
+* Number of accelerators per node
+* Accelerator model name
+
+This package can be invoked as a standalone console script or can be invoked from within
+an application to gather and generate a system environment report.
+
+The module can be invoked by using the entrypoint alias:
+
+.. code-block::
+
+    $ composer_collect_env
+
+Or manually as a standalone script:
+
+.. code-block::
+
+    $ python composer/utils/collect_env.py
+
+To generate a system report from within a user application see :func:`print_env`.
+
+A custom excepthook wrapper is also provided which extends the original :func:`sys.excepthook`
+to automatically collect system information when an exception is raised.
+
+To override the original :func:`sys.excepthook` see :func:`configure_excepthook`.
+
+To restore the original :func:`sys.excepthook` see :func:`restore_excepthook`.
+"""
+
+import sys
+import time
+from typing import NamedTuple, TextIO
+
+import cpuinfo
+import psutil
+
+from composer.utils.import_helpers import import_object
+
+__all__ = ["configure_excepthook", "print_env", "restore_excepthook"]
+
+# Check if PyTorch is installed
+try:
+    import torch.utils.collect_env as torchenv
+    from torch.cuda import device_count as cuda_device_count
+    from torch.cuda import get_device_name as accel_device_name
+    from torch.cuda import is_available as cuda_available
+    TORCH_AVAILABLE = True
+except (ImportError,):
+    TORCH_AVAILABLE = False
+
+# Check if Composer is installed
+try:
+    import composer
+    from composer.utils import dist
+    COMPOSER_AVAILABLE = True
+except (ImportError,):
+    COMPOSER_AVAILABLE = False
+
+# Check if we're running in a notebook
+try:
+    __IPYTHON__  #type: ignore
+    IPYTHON_AVAILABLE = True
+except (NameError,):
+    IPYTHON_AVAILABLE = False
+
+if IPYTHON_AVAILABLE:
+    get_ipython = import_object('IPython:get_ipython')
+    nb = get_ipython()
+
+# Place to keep track of the original excepthook
+_orig_excepthook = None
+
+
+# Same convention as Torch collect_env, create a namedtuple to track collected fields
+class ComposerEnv(NamedTuple):
```

Comment:
nit: Is this named tuple needed? It seems to be used only for making the format string...instead could the format string args be passed in directly?

Reply:
Not really, was originally sticking to the PyTorch `collect_env` convention.  In their case they set the fields to the raw return values of all the `get_*` functions and then use a set of common formatting functions to format the strings it in a function that pretties the string so this bifurcation is more convenient.  In our case for Composer the `get_*` functions output the desired format so it's not really helping.  

We may have a similar pretty function in the future to post process the strings in the future depending on what additional env information we collect and how we collect it (e.g., output of `pip freeze`, exec other sys utilities).  Could just make the change then but given this isn't really a functional or performance issue and more work to change at this point I'm going keep it as is.